File: web/src/types/agui/payloads.ts
import type { MessageRoleValue } from './event-types'
import type { JsonPatchOperation } from './json-patch'

export interface RunStartedPayload {
  threadId: string
  runId: string
  input?: unknown
}

export interface RunFinishedPayload {
  threadId: string
  runId: string
  result?: unknown
}

export interface RunErrorPayload {
  message: string
  code?: string
}

export interface StepStartedPayload {
  stepName: string
}

export interface StepFinishedPayload {
  stepName: string
}

export interface TextMessageStartPayload {
  messageId: string
  role: MessageRoleValue
}

export interface TextMessageContentPayload {
  messageId: string
  delta: string
}

export interface TextMessageEndPayload {
  messageId: string
}

export interface ToolCallStartPayload {
  toolCallId: string
  toolCallName: string
  parentMessageId?: string
}

export interface ToolCallArgsPayload {
  toolCallId: string
  delta: string
}

export interface ToolCallEndPayload {
  toolCallId: string
}

export interface ToolCallResultPayload {
  messageId: string
  toolCallId: string
  content: unknown
  role: MessageRoleValue
}

export interface StateSnapshotPayload {
  snapshot: unknown
}

export interface StateDeltaPayload {
  delta: JsonPatchOperation[]
}

export interface MessagesSnapshotPayload {
  messages: unknown[]
}

export interface CustomPayload {
  name: string
  value: unknown
}

export interface ArtifactCustomValue {
  artifact: unknown
  taskId: string
  contextId: string
}

export interface ExecutionStepCustomValue {
  step: {
    stepId: string
    taskId: string
    status: 'pending' | 'in_progress' | 'completed' | 'failed'
    startedAt: string
    completedAt?: string
    durationMs?: number
    errorMessage?: string
    content: {
      type: 'understanding' | 'planning' | 'skill_usage' | 'tool_execution' | 'completion'
      reasoning?: string
      planned_tools?: Array<{ tool_name: string; arguments: unknown }>
      skill_id?: string
      skill_name?: string
      tool_name?: string
      tool_arguments?: unknown
      tool_result?: unknown
    }
  }
  contextId: string
}

export interface SkillLoadedCustomValue {
  skillId: string
  skillName: string
  description?: string
  taskId?: string
}

export interface RunStartedCustomValue {
  task: {
    id: string
    contextId: string
    status: {
      state: string
      message?: string | null
      timestamp: string
    }
    history: Array<{
      role: string
      parts: Array<{ kind: string; text?: string }>
      messageId: string
      taskId: string | null
      contextId: string
      kind: string
      metadata?: unknown | null
      extensions?: unknown | null
      referenceTaskIds?: string[] | null
    }>
    artifacts?: unknown | null
    metadata?: {
      task_type?: string
      agent_name?: string
      created_at?: string
      [key: string]: unknown
    } | null
    kind: string
  }
  threadId: string
  runId: string
}

export interface TaskCompletedCustomValue {
  task: {
    id: string
    contextId: string
    status: {
      state: string
      message?: string | null
      timestamp: string
    }
    history: Array<{
      role: string
      parts: Array<{ kind: string; text?: string }>
      messageId: string
      taskId: string
      contextId: string
      kind: string
      metadata?: unknown | null
      extensions?: unknown | null
      referenceTaskIds?: string[] | null
    }>
    artifacts?: unknown | null
    metadata?: {
      task_type?: string
      agent_name?: string
      created_at?: string
      updated_at?: string
      started_at?: string
      completed_at?: string
      execution_time_ms?: number
      executionSteps?: Array<{
        stepId: string
        taskId: string
        status: string
        startedAt: string
        completedAt?: string
        durationMs?: number
        content: {
          type: string
        }
      }>
      [key: string]: unknown
    } | null
    kind: string
  }
  artifacts?: unknown[] | null
  executionSteps?: Array<{
    stepId: string
    taskId: string
    status: string
    startedAt: string
    completedAt?: string
    durationMs?: number
    content: {
      type: string
    }
  }>
}

File: crates/shared/models/src/agui/payloads.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use systemprompt_identifiers::{AiToolCallId, ContextId, MessageId, TaskId};

use super::JsonPatchOperation;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum MessageRole {
    User,
    Assistant,
    System,
    Tool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunStartedPayload {
    pub thread_id: ContextId,
    pub run_id: TaskId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunFinishedPayload {
    pub thread_id: ContextId,
    pub run_id: TaskId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunErrorPayload {
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub code: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StepStartedPayload {
    pub step_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StepFinishedPayload {
    pub step_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TextMessageStartPayload {
    pub message_id: MessageId,
    pub role: MessageRole,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TextMessageContentPayload {
    pub message_id: MessageId,
    pub delta: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TextMessageEndPayload {
    pub message_id: MessageId,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolCallStartPayload {
    pub tool_call_id: AiToolCallId,
    pub tool_call_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_message_id: Option<MessageId>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolCallArgsPayload {
    pub tool_call_id: AiToolCallId,
    pub delta: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolCallEndPayload {
    pub tool_call_id: AiToolCallId,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolCallResultPayload {
    pub message_id: MessageId,
    pub tool_call_id: AiToolCallId,
    pub content: Value,
    pub role: MessageRole,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StateSnapshotPayload {
    pub snapshot: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StateDeltaPayload {
    pub delta: Vec<JsonPatchOperation>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessagesSnapshotPayload {
    pub messages: Vec<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArtifactCustomPayload {
    pub artifact: crate::a2a::Artifact,
    pub task_id: TaskId,
    pub context_id: ContextId,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExecutionStepCustomPayload {
    pub step: crate::execution::ExecutionStep,
    pub context_id: ContextId,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillLoadedCustomPayload {
    pub skill_id: systemprompt_identifiers::SkillId,
    pub skill_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub task_id: Option<TaskId>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GenericCustomPayload {
    pub name: String,
    pub value: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "name", content = "value", rename_all = "snake_case")]
pub enum CustomPayload {
    Artifact(Box<ArtifactCustomPayload>),
    ExecutionStep(Box<ExecutionStepCustomPayload>),
    SkillLoaded(SkillLoadedCustomPayload),
    #[serde(untagged)]
    Generic(GenericCustomPayload),
}
